src/main/java/hexgui/gui/FileDialogs.java

```diff
@@ -0,0 +1,190 @@
+  private FileDialogs() {}
+
+  /**
+   * Find first parent that is a Frame.
+   *
+   * @return null If no such parent.
+   */
+  private static Frame findParentFrame(Component component) {
+    while (component != null)
+      if (component instanceof Frame) return (Frame) component;
+      else component = component.getParent();
+    return null;
+  }
+
+  private static File showFileChooser(
+      Component parent, Type type, File lastFile, boolean setSgfFilter, String title) {
+    // Use native dialogs for some platforms. but not for type select
+    // There is no native dialog for select
+    if (NATIVE_DIALOGS && type != Type.FILE_SELECT) {
+      Frame frame = findParentFrame(parent);
+      return showFileChooserAWT(frame, type, title);
+    }
+    return showFileChooserSwing(parent, type, lastFile, setSgfFilter, title);
+  }
+
+  private static File showFileChooserSave(
+      Component parent,
+      File lastFile,
+      boolean setSgfFilter,
+      String title,
+      MessageDialogs messageDialogs) {
+    File file = showFileChooser(parent, Type.FILE_SAVE, lastFile, setSgfFilter, title);
+    if (NATIVE_DIALOGS)
+      // Overwrite warning is already part of FileDialog
+      return file;
+    while (file != null) {
+      if (file.exists()) {
+        String mainMessage = MessageFormat.format("Replace file \"{0}\"", file.getName());
+        String optionalMessage = "If you overwrite the file, the previous version will be lost.";
+        if (!messageDialogs.showQuestion(parent, mainMessage, optionalMessage, "Replace", true)) {
+          file = showFileChooser(parent, Type.FILE_SAVE, lastFile, setSgfFilter, title);
+          continue;
+        }
+      }
+      break;
+    }
+    return file;
+  }
+
+  private static File showFileChooserAWT(Frame parent, Type type, String title) {
+    FileDialog dialog = new FileDialog(parent);
+    if (title == null) {
+      switch (type) {
+        case FILE_OPEN:
+          title = "Open";
+          break;
+        case FILE_SAVE:
+          title = "Save";
+          break;
+        default:
+          assert false;
+      }
+    }
+    dialog.setTitle(title);
+    int mode = FileDialog.LOAD;
+    if (type == Type.FILE_SAVE) mode = FileDialog.SAVE;
+    dialog.setMode(mode);
+    /* Commented out, because there is no way to change the filter by the
+       user (at least not on Linux)
+    if (setSgfFilter)
+        dialog.setFilenameFilter(new FilenameFilter() {
+                public boolean accept(File dir, String name)
+                {
+                    return name.toLowerCase().endsWith("sgf");
+                }
+            });
+    */
+    // dialog.setLocationRelativeTo(parent); // Java <= 1.4
+    dialog.setLocationByPlatform(true);
+    dialog.setVisible(true);
+    if (dialog.getFile() == null) return null;
+    return new File(dialog.getDirectory(), dialog.getFile());
+  }
+
+  private static File showFileChooserSwing(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 174-185 into a separate method.